### ISSUE TITLE
Memory allocation when transcribing

### ISSUE THREAD

**Issue by DiegoJota** [NVIDIA/NeMo#2765]
Hi

I'm building a communication API for transcription with NeMo and I ran into a critical infinite memory allocation problem.
When using a lib that monitors memory allocation, I noticed that the "transcribe" method used on top of a restored model is allocating memory every time when transcribing a new file.
My API basically works like this:
When starting, I load the .nemo model creating the asr_model, then open a listener and keep receiving hundreds of files on the network, carrying out their individual transcription on top of that same nemo_asr object.
At this point the problem comes up, because every time I call the transcribe method it keeps allocating more and more memory... I left it running for a night and it allocated 50GB of ram (considering a small stream of audios to transcribe)
Is there an alternative so I can work around this problem? I don't think it's perfect to have to create a new nemo_asr object every time I transcribe, so I keep it in this format.
Here are two examples below:
![Ev1](https://user-images.githubusercontent.com/49459897/131869414-b711961f-0640-4f73-8078-1e15d6832787.PNG)
![Ev2](https://user-images.githubusercontent.com/49459897/131869432-3e259a63-a8c2-4165-bdb2-88ada3ffc860.PNG)

SO: Ubuntu 18.04
Nemo: 1.2.0

**Comment by titu1994** [comment]
Try deleting the logits once you are done with it and calling gc explicitly. Transcribe on gpu does not explode memory, I've tried it with 100k files without issue. Possibly the returned tensor is used somewhere in the code in a loop and not released ?

**Comment by titu1994** [comment]
As to why it's allocating memory at all - you have logprobs=True which means it needs to return the output rank 3 tensor which is going to cost some memory

**Comment by titu1994** [comment]
Also, NeMo 1.3 is released so you could try that, but I don't recall any changes to transcribe which would affect memory

**Comment by DiegoJota** [comment]
I already do the explicit cleanup of all objects that are returned. But without success
```
del probs, non_speech, transcript, spaces, labels, words, logits
gc.collect()
```
When using sys.getsizeof() I see that logits and probs are less than 0.5mb. That is, there are still 22 mb lost that I don't know where they might be, which are always allocated when calling "transcribe"

**Comment by DiegoJota** [comment]
And taking advantage of this, I'm just referring to the transcribe but the diarizer and beam_search processes are also having the same problem.
![image](https://user-images.githubusercontent.com/49459897/131921366-ab434544-080d-44ff-9e75-6a361c972e2b.png)
![image](https://user-images.githubusercontent.com/49459897/131921568-9eef8d28-77ca-47a3-ad01-35be9da1df02.png)

**Comment by DiegoJota** [comment]
As it is an API, every audio that arrives I start a Thread because the processing cannot be synchronous as it can be quite time consuming. I suspect the Thread is holding the allocation, but with no success so far.

**Comment by titu1994** [comment]
Have you tried freezing the model with model.freeze() after instantiation? I think transcribe does this automatically though. This doesn't occur in Nemo normal execution and multinode execution. 

Possibly an artifact of cpu + multi threading + pytorch. Easiest thing would be to use a multiprocess frameworks such as gunicorn or something to manage stuff properly
